Introduce PerAxis Quantized constraint for StableHLO Quantized OPs

abhigunj opened the pull request:
StableHLO OPs supporting Quantization can be categorized into following three types
a. Only PerTensor Quantized Tensors
b. Only PerAxis Quantized Tensors (no OP in this category for now)
c. Both PerTensor and PerAxis Quantized Tensors
PerAxis constraint from the PR allow only PerTensorQuantized Tensor inputs to type (a) OPs  and don't allow  PerAxis Quantized Tensors. 
Also,
Added negative test cases to validate this behavior
Added positive test cases for PerTensor , PerAxis Quantized Tensor support


Excluded OPs 
* Deprecated :`BroadCast` and `Call`
* Not Specced :  
 `dynamic_iota`, `create_token`, `dynamic_broadcast_in_dim`, `cross-replica-sum`, `einsum`, `unary_einsum`, `dynamic_reshape`, `set_dimension_size`, `trace`, `return`, `torch_index_select`, `real_dynamic_slice`, `dynamic_pad`, `dynamic_gather`, `dynamic_conv`, `dynamic_reshape_shape`, `cstr_reshapable`

GleasonK commented on stablehlo/tests/ops_stablehlo_quantized.mlir:
```diff
@@ -0,0 +1,489 @@
+    indices_are_sorted = false
+  } : (tensor<*x!quant.uniform<i8:f32:0, {0.1:-30}>>, tensor<1x5x2xi32>) -> tensor<8x?x7x1x6x1x?x!quant.uniform<i8:f32:0, {0.1:-30}>>
+  func.return %res : tensor<8x?x7x1x6x1x?x!quant.uniform<i8:f32:0, {0.1:-30}>>
+}
```
newline at EOF

GleasonK replied:
Why didn't "Lint / whitespace-checks" catch this, maybe open a ticket. (Not looking at *.mlir files maybe?)

abhigunj replied:
Tracking this at https://github.com/openxla/stablehlo/issues/2010.

GleasonK commented on stablehlo/tests/ops_stablehlo_quantized.mlir:
```diff
@@ -0,0 +1,489 @@
+
+func.func @negative_floor_quantization(%arg0: tensor<1x2x2x!quant.uniform<i8:f32:0, {0.1:-30}>>){
+  // expected-error@+1 {{operand #0 must be tensor of f8E4M3B11FNUZ type or f8E4M3FN type or f8E4M3FNUZ type or f8E5M2 type or f8E5M2FNUZ type or 16-bit float or 32-bit float or 64-bit float or bfloat16 type or 4/8/16/32-bit uniform quantized signed integer or 4/8/16/32-bit uniform quantized unsigned integer values, but got 'tensor<1x2x2x!quant.uniform<i8:f32:0, {1.000000e-01:-30}>>'}}
+  %is_finite = "stablehlo.is_finite"(%arg0) {} : (tensor<1x2x2x!quant.uniform<i8:f32:0, {0.1:-30}>>) -> tensor<1x2x2xi1>
```
Need to consider compatibility around these `FP --> FP or Quant` operations. Can use something similar to ReduceOp's recent compat changes

GleasonK replied:
This is a compatibility exception, as quantization was previously unspecced but had limited supported, and unspecced features were exempt from full compatibility guarantees.

If this were to have a broader compat impact we could offser some support, but I don't think anyone is relying on these in serialization yet. (JAX is not emitting them for example).

GleasonK commented:
Delegating my review to @sdasgup3. I took a scan over the code / tests and from a high level everything LGTM.

sdasgup3 commented on stablehlo/dialect/StablehloOps.td:
```diff
@@ -2109,11 +2110,11 @@ def StableHLO_ConvolutionOp : StableHLO_Op<"convolution", [Pure]> {
     (ins
-       HLO_Tensor:$lhs,
-       HLO_Tensor:$rhs),
+       HLO_TensorOrPerAxisQuantized:$lhs,
```
`lhs` is per-tensor.

sdasgup3 reviewed (commented):
Thanks again for the work. Here is some of my initial reviews which I want to publish soon. I need to check the test files in my follow up review. 

Here is the list of ops which are unhandled:


- constant op: is current using `HLO_StaticShapeTensor` which per this change has per-tensor output type. It should be both.
 - `iota` uses `HLO_StaticShapeIntFpOrComplexTensor` which is not updated. iota is expected to support `per-tensor`
 - `infeed`: Accepts both per-tensor and per-axis.
 - `recv`: should take both per-tensor and per-axis. 
 - `If`, `Case`, `While`: The output supports both.
 - `GetTupleElement` / `Tuple`:  supports both.
 - `broadcast`: update the type coinstraints treating the in/out types similar to `brodcast_in_dim`
 - `custom-call`: both - `fft`: fix the result type as per spec
- `transpose`: both are supported. - 

Nit:
 -  `dynamic_iota`, `create_token`, `dynamic_broadcast_in_dim`, `cross-replica-sum`, `einsum`, `unary_einsum`, `dynamic_reshape`, `set_dimension_size`, `trace`, `return`, `torch_index_select`, `real_dynamic_slice`, `dynamic_pad`, `dynamic_gather`, `dynamic_conv`, `dynamic_reshape_shape`, `cstr_reshapable`: mention in the description that we are excluding the stablehlo ops not specced  and enumerate only these. The idea is to clearly mention  which ops in .td file are not specced. 


As follow up PRs:
 - collective_broadcast should also support per-tensor type like other distribution ops.  
  - get_dimension_size: fix the spec

sdasgup3 commented on stablehlo/dialect/Base.td:
```diff
@@ -174,6 +214,10 @@ def HLO_StaticDimensionTensor : RankedTensorOf<[HLO_DimensionValue], [HasStaticS
       HLO_Float, HLO_Pred, HLO_Int, HLO_Complex, HLO_QuantizedInt]>;
 
+// To address broadcast_in_dimm, reshape OP test failures
+def HLO_StaticShapeOrQuantizedPerAxis : StaticShapeTensorOf<[
```
`HLO_StaticShapeOrQuantizedPerAxis` -> `HLO_StaticShapeTensorOrPerAxisQuantizedtensor`

abhigunj replied:
`HLO_StaticShapeTensor` is used for defining `HLO_StaticShapeTensorOrToken` and similar other defs. 

Removed `HLO_StaticShapeTensorOrToken`

sdasgup3 commented on stablehlo/tests/ops_stablehlo_quantized.mlir:
```diff
@@ -0,0 +1,823 @@
+}
+
+
+// Negative Tests for StableHLO OPs supporting PerTensorQuantizedTensors but not PerAxisQuantizedTensors
```
supporting per-tensor quantization only (not per-axis)

sdasgup3 replied:
For consistency put the comment after the split (same as in line 39)

sdasgup3 commented on stablehlo/tests/ops_stablehlo_quantized.mlir:
```diff
@@ -0,0 +1,823 @@
+}
+
+// -----
+// Tests for StableHLO OPs supporting PerTensorQuantizedTensors
```
Positive tests for ops supporting only per-tensor quantization

sdasgup3 commented on stablehlo/tests/ops_stablehlo_quantized.mlir:
```diff
@@ -0,0 +1,823 @@
+  %arg3: tensor<2x4x!quant.uniform<i8:f32, 1.0:17>>,
+  %token0: !stablehlo.token) {
+  %abs = "stablehlo.abs"(%arg0) : (tensor<1x2x2x!quant.uniform<i8:f32, 1.0:17>>) -> tensor<1x2x2x!quant.uniform<i8:f32, 1.0:17>>
+  %add = "stablehlo.add"(%arg0, %arg1) : (tensor<1x2x2x!quant.uniform<i8:f32, 1.0:17>>, tensor<1x2x2x!quant.uniform<i8:f32, 1.0:17>>) -> tensor<1x2x2x!quant.uniform<i8:f32, 1.0:17>>
```
add does not fall into this bucket anymore

sdasgup3 commented on stablehlo/tests/ops_stablehlo_quantized.mlir:
```diff
@@ -0,0 +1,823 @@
+
+// -----
+
+func.func @negative_abs_quantization(%arg0: tensor<1x2x2x!quant.uniform<i8:f32:0, {0.1:-30}>>){
```
update the func name to `negative_abs_per_axis_quantization`? 
and below

sdasgup3 commented on stablehlo/tests/ops_stablehlo_quantized.mlir:
```diff
@@ -0,0 +1,823 @@
+  func.return
+}
+
+func.func @batch_norm_grad_quantization(%input: tensor<2x2x2x2x!quant.uniform<i8:f32, 1.0:17>>, %scale: tensor<2x!quant.uniform<i8:f32, 1.0:17>>, %mean: tensor<2x!quant.uniform<i8:f32, 1.0:17>>, %variance: tensor<2x!quant.uniform<i8:f32, 1.0:17>>, %grad_output: tensor<2x2x2x2x!quant.uniform<i8:f32, 1.0:17>>) -> tensor<2x2x2x2x!quant.uniform<i8:f32, 1.0:17>> {
```
`batch_norm_grad_per_tensor_quantization`?
and below

abhigunj replied:
Tests are categorized as:
* Tests for per-axis Q OPs
* Tests for per-tensor Q OPs ( few OP also supports per-axis Q)
* negative Tests (for both per-axis and per-tensor Q OPs)
updated the comments to make it clearer.

with this, will keep the naming as it is.

abhigunj replied:
added per-axis and per-tensor in test names

sdasgup3 commented on stablehlo/tests/ops_stablehlo_quantized.mlir:
```diff
@@ -0,0 +1,823 @@
+// -----
+// Tests for StableHLO OPs supporting PerTensorQuantizedTensors
+
+func.func @quantization_supported_ops(
```
`ops_with_per_tensor_quantization`

abhigunj commented:
> Thanks again for the work. Here is some of my initial reviews which I want to publish soon. I need to check the test files in my follow up review.
> 
> Here is the remaining list of ops which are uncovered and the following support need to be added.
> 
> * `constant` : is current using `HLO_StaticShapeTensor` which per this change has per-tensor output type. It should be both.
> * `iota` uses `HLO_StaticShapeIntFpOrComplexTensor` which is not updated. iota is expected to support `per-tensor`
> * `infeed`: Accepts both per-tensor and per-axis.
> * `recv`: should take both per-tensor and per-axis.
> * `If`, `Case`, `While`: The output supports both.
> * `GetTupleElement` / `Tuple`:  supports both.
> * `broadcast`: update the type coinstraints treating the in/out types similar to `brodcast_in_dim`
> * `custom-call`: both - `fft`: fix the result type as per spec.

Thanks for the thorough review, these OPs fall into result Quantized category, which I incorrectly ignored during audit. Updated the OP def. Updated the audit sheet.

> * `transpose`: both are supported.

Already taken care of?

> 
> Nit:  -  `dynamic_iota`, `create_token`, `dynamic_broadcast_in_dim`, `cross-replica-sum`, `einsum`, `unary_einsum`, `dynamic_reshape`, `set_dimension_size`, `trace`, `return`, `torch_index_select`, `real_dynamic_slice`, `dynamic_pad`, `dynamic_gather`, `dynamic_conv`, `dynamic_reshape_shape`, `cstr_reshapable`: Let's mention in the description that we are excluding the above stablehlo ops as they are not specced.
> 

Done

> As follow up PRs:
> 
> * collective_broadcast should also support per-tensor type like other distribution ops.
> * get_dimension_size: fix the spec
>   Please free to open PRs for the above items.

Yes, will create separate PR as it involves changes to the spec.

abhigunj commented on stablehlo/tests/ops_stablehlo_quantized.mlir:
```diff
@@ -0,0 +1,823 @@
+// Tests for StableHLO OPs supporting PerAxisQuantizedTensors
+
+// CHECK-LABEL: @per_axis_quantized_ops
+func.func @per_axis_quantized_ops(
```
These OPs also support per-tensor quantized inputs

sdasgup3 commented on stablehlo/dialect/StablehloOps.td:
```diff
@@ -1862,11 +1863,11 @@ def StableHLO_BroadcastInDimOp : StableHLO_Op<"broadcast_in_dim",
   }];
   let arguments = (ins
-    HLO_Tensor:$operand /*broadcast_in_dim_i1*/,
+    HLO_TensorOrPerAxisQuantizedTensor:$operand /*broadcast_in_dim_i1*/,
```
Can we make similar changes in `broadcast` as well?

abhigunj replied:
Done, I left it untouched after reading a comment -  on its way out of StableHLO

sdasgup3 commented on stablehlo/dialect/Base.td:
```diff
@@ -174,13 +218,19 @@ def HLO_StaticDimensionTensor : RankedTensorOf<[HLO_DimensionValue], [HasStaticS
 def HLO_StaticShapeTensor : StaticShapeTensorOf<[
```
remove this as it will not be used anywhere?

abhigunj replied:
it is being used for def HLO_StaticShapeTensorOrPerAxisQuantizedTensorOrToken

sdasgup3 reviewed (approved):
lgtm with some minor comments.

GleasonK commented on stablehlo/dialect/Base.td:
```diff
@@ -45,36 +45,46 @@ def HLO_Complex : Complex<AnyTypeOf<[F32, F64]>>;
 
 // TODO(b/230381284): Upstream width-specific uniform quantized element types.
 class StableHLO_UniformQuantizedSignedInt<int width>
```
`StableHLO_UniformPerTensorQuantizedSignedInt` ?

abhigunj replied:
Revisited this:  we are using `*Quantized*` to indicate per-tensor quantization and `*QuantizedPerAxis*` to indicate per-axis quantization. this matches with MLIR conventions `quant::UniformedQuantizedType` (per-tensor) and `quant::UniformedPerAxisQuantized` (per-axis).  Will keep `StableHLO_UniformQuantizedSignedInt ` unchanged.

GleasonK commented on stablehlo/dialect/Base.td:
```diff
@@ -146,6 +187,9 @@ def HLO_PredOrIntTensor : TensorOf<[HLO_Pred, HLO_Int]>;
 def HLO_FpOrComplexTensor : TensorOf<[HLO_Float, HLO_Complex]>;
 
+// Any floating-point or quantized tensor types
+def HLO_FpOrQuantizedIntTensor : TensorOf<[HLO_Float, HLO_QuantizedInt]>;
```
I asked this in the T2x PR, but how many FpTensor ops also take quantized types? Don't need exact number, but if the answer is 90% then lets update FpTensor to allow per-tensor quantized ints

abhigunj replied:
8 out of 14 FpTensor OPs take quantized types. Better to keeping Quantized type separate?

abhigunj replied:
revisited: there are 6 ops which doesn't take quantized types, few also take other types like complex, it appear ok to keep `float` and `quant` separate and use `HLO_FpOrQuantizedIntTensor `.

GleasonK commented on stablehlo/dialect/StablehloOps.td:
```diff
@@ -1708,19 +1709,19 @@ def StableHLO_BatchNormGradOp : StableHLO_Op<"batch_norm_grad", [Pure,
-    1DTensorOf<[HLO_Float]>:$mean, /*batch_norm_grad_i3*/
-    1DTensorOf<[HLO_Float]>:$variance, /*batch_norm_grad_i4*/
-    RankedTensorOf<[HLO_Float]>:$grad_output, /*batch_norm_grad_i5*/
+    RankedTensorOf<[HLO_Float, HLO_QuantizedInt]>:$operand, /*batch_norm_grad_i1*/
```
Left on a future CL, but let's add a ticket to make this `FpTensor` once there is no unranked tensor type in StableHLO

abhigunj replied:
for now, prefer not to change it to FpTensor.  Created ticket to review this once all Q verifier changes are done.

abhigunj replied:
revisited: there are 6 ops which doesn't take quantized types, few also take other types like complex, it appear ok to keep float and quant separate and use HLO_FpOrQuantizedIntTensor .

GleasonK reviewed (approved):
I'll think more on naming, but that shouldn't block this from going in since tablegen variable names don't impact the generated code / impl.
